src/blocking.rs

```diff
@@ -0,0 +1,2830 @@
+
+            let url_base = url
+                .strip_suffix(image.tag().unwrap_or("latest"))
+                .expect("The manifest URL always ends with the image tag suffix");
+            let url_by_digest = format!("{}{}", url_base, manifest_hash);
+
+            return Ok(url_by_digest);
+        }
+
+        ret
+    }
+
+    /// Pulls the referrers for the given image filtering by the optionally provided artifact type.
+    pub fn pull_referrers(
+        &mut self,
+        image: &Reference,
+        artifact_type: Option<&str>,
+    ) -> Result<OciImageIndex> {
+        let url = self.to_v2_referrers_url(image, artifact_type)?;
+        debug!("Pulling referrers from {}", url);
+
+        let res = RequestBuilderWrapper::from_client(self, |client| client.get(&url))
+            .apply_accept(MIME_TYPES_DISTRIBUTION_MANIFEST)?
+            .apply_auth(image, RegistryOperation::Pull)?
+            .into_request_builder()
+            .send()?;
+        let status = res.status();
+        let body = res.bytes()?;
+
+        validate_registry_response(status, &body, &url)?;
+        let manifest = serde_json::from_slice(&body)
+            .map_err(|e| OciDistributionError::ManifestParsingError(e.to_string()))?;
+
+        Ok(manifest)
+    }
+
+    fn extract_location_header(
+        &self,
+        image: &Reference,
+        res: reqwest::blocking::Response,
+        expected_status: &reqwest::StatusCode,
+    ) -> Result<String> {
+        debug!(expected_status_code=?expected_status.as_u16(),
+            status_code=?res.status().as_u16(),
+            "extract location header");
+        if res.status().eq(expected_status) {
+            let location_header = res.headers().get("Location");
+            debug!(location=?location_header, "Location header");
+            match location_header {
+                None => Err(OciDistributionError::RegistryNoLocationError),
+                Some(lh) => self.location_header_to_url(image, lh),
+            }
+        } else if res.status().is_success() && expected_status.is_success() {
+            Err(OciDistributionError::SpecViolationError(format!(
+                "Expected HTTP Status {}, got {} instead",
+                expected_status,
+                res.status(),
+            )))
+        } else {
+            let url = res.url().to_string();
+            let code = res.status().as_u16();
+            let message = res.text()?;
+            Err(OciDistributionError::ServerError { url, code, message })
+        }
+    }
+
+    /// Helper function to convert location header to URL
+    ///
+    /// Location may be absolute (containing the protocol and/or hostname), or relative (containing just the URL path)
+    /// Returns a properly formatted absolute URL
+    fn location_header_to_url(
```

Comment:
This method could probably be moved to the `ClientConfig` struct as a `pub(crate)` so it can be reused across the clients